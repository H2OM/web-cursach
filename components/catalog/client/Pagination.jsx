'use client';

import { usePathname, useRouter, useSearchParams } from "next/navigation";


export default function Pagination ({page, max}) {
    const router = useRouter();
    const pathname = usePathname();
    const searchParams = useSearchParams();
    const params = new URLSearchParams(searchParams);
    if(max == 1) return;
    
    return (
        <nav className="Catalog__nav" onClick={({target})=>{
            if(target.classList.contains("Catalog__nav__point") && !target.classList.contains("Catalog__nav__point_select")) {
                params.set('page', target.textContent);
                router.push(pathname+'?'+ params.toString());
            }
        }}>
            <>
                {
                    page > 4 ? 
                        <>
                            {
                                (page-(6 - (max-page))) < 3 ? null : 
                                    <><div className="Catalog__nav__point">1</div>
                                    <div className="Catalog__nav__space">...</div></>
                            }
                            {   
                                max - page >= 5 ? 
                                    <>
                                        <div className="Catalog__nav__point">{page - 2}</div>
                                        <div className="Catalog__nav__point">{page - 1}</div>
                                    </> : 
                                    Array.from({length: (6 - (max - page))}).map((_, i)=>{    //перед
                                        return (
                                            <div className="Catalog__nav__point" key={10+i}>{page-(6 - (max-page))+i}</div>
                                        )
                                    }) 
                            }
                            <div className="Catalog__nav__point Catalog__nav__point_select">{page}</div>
                        </> 
                        :
                        Array.from({length: page}).map((_, i)=>{ 
                            return (
                                <div className={"Catalog__nav__point" + ((page == i+1) ? " Catalog__nav__point_select" : "")} key={20+i}>{i+1}</div>
                            )
                        }) 
                }
                {
                    Array.from({length: ((max - page) < 7 ? max - page : 5)}).map((_, i)=>{ //после
                                                
                        return (
                            <div className="Catalog__nav__point" key={30+i}>{page+i+1}</div>
                        )
                    }) 
                }
                {
                    max - page >= 7 ?
                        <>
                            <div className="Catalog__nav__space">...</div>
                            <div className="Catalog__nav__point">{max}</div>
                        </>
                            : null
                }
            </>
        </nav>
    )
}