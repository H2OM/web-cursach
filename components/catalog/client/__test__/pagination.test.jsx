import { usePathname, useSearchParams, useRouter } from 'next/navigation';
import { render, screen } from "@testing-library/react";
import Pagination from '../Pagination';
jest.mock('next/navigation');
useRouter.mockReturnValue({
    push: jest.fn(),
}); 
usePathname.mockReturnValue("localhost");
useSearchParams.mockReturnValue({});

const urlparams = new URLSearchParams();
global.URLSearchParams = jest.fn(x => (urlparams));


describe('pagination test', ()=>{
    it('render test start', ()=>{
        const {container} = render(<Pagination page={1} max={40}/>);
        const dots = screen.getAllByText('...');

        expect(dots.length).toBe(1);
        expect(screen.getByText(1)).toBeInTheDocument();
        
        Array.from({length: 5}).forEach((_,i)=>{
            expect(screen.getByText(1+i+1)).toBeInTheDocument();
        });

        expect(screen.getByText(40)).toBeInTheDocument();

        expect(container).toMatchSnapshot();
    });
    it('render test middle', ()=>{
        const {container} = render(<Pagination page={8} max={40}/>);
        const dots = screen.getAllByText('...');


        expect(screen.getByText('1')).toBeInTheDocument();
        expect(dots.length).toBe(2);
        expect(screen.getByText(8)).toBeInTheDocument();
        expect(screen.getByText(8-1)).toBeInTheDocument();
        expect(screen.getByText(8-2)).toBeInTheDocument();
        
        Array.from({length: 5}).forEach((_,i)=>{
            expect(screen.getByText(8+i+1)).toBeInTheDocument();
        });
        expect(screen.getByText(40)).toBeInTheDocument();

        expect(container).toMatchSnapshot();
    });
    it('render test end', ()=>{
        const {container} = render(<Pagination page={40} max={40}/>);
        const dots = screen.getAllByText('...');

        expect(dots.length).toBe(1);
        expect(screen.getByText(1)).toBeInTheDocument();
                
        Array.from({length: 6}).forEach((_,i)=>{
            expect(screen.getByText(40-6+i)).toBeInTheDocument();
        });

        expect(screen.getByText(40)).toBeInTheDocument();

        expect(container).toMatchSnapshot();
    });
});