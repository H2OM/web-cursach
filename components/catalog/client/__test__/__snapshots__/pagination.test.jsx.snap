// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pagination test render test 1`] = `
<div>
  <nav
    class="Catalog__nav"
  >
    <div
      class="Catalog__nav__point"
    >
      1
    </div>
    <div
      class="Catalog__nav__space"
    >
      ...
    </div>
    <div
      class="Catalog__nav__point"
    >
      6
    </div>
    <div
      class="Catalog__nav__point"
    >
      7
    </div>
    <div
      class="Catalog__nav__point Catalog__nav__point_select"
    >
      8
    </div>
    <div
      class="Catalog__nav__point"
    >
      9
    </div>
    <div
      class="Catalog__nav__point"
    >
      10
    </div>
    <div
      class="Catalog__nav__point"
    >
      11
    </div>
    <div
      class="Catalog__nav__point"
    >
      12
    </div>
    <div
      class="Catalog__nav__point"
    >
      13
    </div>
    <div
      class="Catalog__nav__space"
    >
      ...
    </div>
    <div
      class="Catalog__nav__point"
    >
      40
    </div>
  </nav>
</div>
`;

exports[`pagination test render test end 1`] = `
<div>
  <nav
    class="Catalog__nav"
  >
    <div
      class="Catalog__nav__point"
    >
      1
    </div>
    <div
      class="Catalog__nav__space"
    >
      ...
    </div>
    <div
      class="Catalog__nav__point"
    >
      34
    </div>
    <div
      class="Catalog__nav__point"
    >
      35
    </div>
    <div
      class="Catalog__nav__point"
    >
      36
    </div>
    <div
      class="Catalog__nav__point"
    >
      37
    </div>
    <div
      class="Catalog__nav__point"
    >
      38
    </div>
    <div
      class="Catalog__nav__point"
    >
      39
    </div>
    <div
      class="Catalog__nav__point Catalog__nav__point_select"
    >
      40
    </div>
  </nav>
</div>
`;

exports[`pagination test render test middle 1`] = `
<div>
  <nav
    class="Catalog__nav"
  >
    <div
      class="Catalog__nav__point"
    >
      1
    </div>
    <div
      class="Catalog__nav__space"
    >
      ...
    </div>
    <div
      class="Catalog__nav__point"
    >
      6
    </div>
    <div
      class="Catalog__nav__point"
    >
      7
    </div>
    <div
      class="Catalog__nav__point Catalog__nav__point_select"
    >
      8
    </div>
    <div
      class="Catalog__nav__point"
    >
      9
    </div>
    <div
      class="Catalog__nav__point"
    >
      10
    </div>
    <div
      class="Catalog__nav__point"
    >
      11
    </div>
    <div
      class="Catalog__nav__point"
    >
      12
    </div>
    <div
      class="Catalog__nav__point"
    >
      13
    </div>
    <div
      class="Catalog__nav__space"
    >
      ...
    </div>
    <div
      class="Catalog__nav__point"
    >
      40
    </div>
  </nav>
</div>
`;

exports[`pagination test render test start 1`] = `
<div>
  <nav
    class="Catalog__nav"
  >
    <div
      class="Catalog__nav__point Catalog__nav__point_select"
    >
      1
    </div>
    <div
      class="Catalog__nav__point"
    >
      2
    </div>
    <div
      class="Catalog__nav__point"
    >
      3
    </div>
    <div
      class="Catalog__nav__point"
    >
      4
    </div>
    <div
      class="Catalog__nav__point"
    >
      5
    </div>
    <div
      class="Catalog__nav__point"
    >
      6
    </div>
    <div
      class="Catalog__nav__space"
    >
      ...
    </div>
    <div
      class="Catalog__nav__point"
    >
      40
    </div>
  </nav>
</div>
`;
